### ISSUE TITLE
Add concept of "optional" parameter

### ISSUE THREAD

**Issue by dhermes** [wntrblm/nox#62]
For example, if I wanted to (as a `nox` user) impl. my own support for arch (as in #60), then I'd want to do something like:

```
$ nox -s "unit(py='2.7', arch='32')"
```

but I **wouldn't** want

```
$ nox -s "unit(py='2.7')"
```

to run over **all choices** of `arch`, I would just want `arch` to be ignored (`arch` would only be relevant on OS X).

Since `nox` has a lot of `pytest`-like interfaces, I [threw together][1] an example of how `pytest` deals with optional test cases with `pytest.mark.parametrize`. (Isn't a perfect mapping.)

----

**NOTE**: @jonparrott and I discussed this earlier this week, I just never filed an issue

[1]: https://gist.github.com/dhermes/fc2c198ca2ebece79d0a05d20a5106b4

**Comment by dhermes** [comment]
@jonparrott Optional parameters are useful **independent** of specifying architecture as in #60 (and discussed in #63).

ISTM that implementing a way to specify or pass-through the architecture would be **much** easier if we had optional parameters in `nox`.

So tackling this would both add a nice feature and enable a simpler impl. of the architecture feature as well. WYDT?

**Comment by theacodes** [comment]
@dhermes do you still think this is relevant?

**Comment by dhermes** [comment]
Yes I do still think it's relevant, since there are certain parameter combinations that only make since in restricted environments (e.g. in a container, on Windows, etc.)

**Comment by theacodes** [comment]
The blocker on this issue is figuring out exactly how this would work and how we would invoke it using `-s` and `-k`.

**Comment by alex** [comment]
I believe this is similar to a need I have, though the way I thought of it was: a way to specify a default value for a param.

Usage:

Given a `noxfile.py` like:

```py
@nox.session
@nox.parametrize("ssh", [False, True])
@nox.parametrize("randomorder", [False, True])
@nox.parametrize("coverage", [False, True])
def tests(
    session: nox.Session, ssh: bool, randomorder: bool, coverage: bool
) -> None:
    ...
```

Right now you'll get 8 builds if you run `nox -e 'tests' -l`. Ideally we'd be able to write:

```py
@nox.session
@nox.parametrize("ssh", [False, True], default=False)
@nox.parametrize("randomorder", [False, True], default=False)
@nox.parametrize("coverage", [False, True], default=True)
def tests(
    session: nox.Session, ssh: bool, randomorder: bool, coverage: bool
) -> None:
    ...
```

And then `nox -e 'tests' -l` would show 1 build, and if I wanted the non-default parameters I'd have to explicitly run `nox -e 'tests(ssh=True)`.


This doesn't have to be attached to `nox.parametrize`, it just seemed like a natural way to extend the existing model.
